app/src/test/java/com/nervousfish/nervousfish/modules/database/DatabaseTest.java

```diff
@@ -0,0 +1,161 @@
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.nervousfish.nervousfish.BaseTest.accessConstructor;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DatabaseTest {
+
+    private final static String CONTACTS_PATH = "temp_contacts.json";
+    private final static String USERDATA_PATH = "temp_userdata.json";
+
+    private IServiceLocatorCreator serviceLocatorCreator = mock(IServiceLocatorCreator.class);
+    private IServiceLocator serviceLocator = mock(IServiceLocator.class);
+    private IConstants constants = mock(IConstants.class);
+
+    private IDatabase database;
+
+    @Before
+    public void setup() {
+        when(serviceLocatorCreator.getServiceLocator()).thenReturn(serviceLocator);
+        when(serviceLocator.getConstants()).thenReturn(constants);
+        when(constants.getDatabaseContactsPath()).thenReturn(CONTACTS_PATH);
+        when(constants.getDatabaseUserdataPath()).thenReturn(USERDATA_PATH);
+
+        this.database = (GsonDatabaseAdapter) accessConstructor(GsonDatabaseAdapter.class, serviceLocatorCreator);
+    }
+
+    @After
+    public void tearDown() {
+        File contactsFile = new File(CONTACTS_PATH);
+        contactsFile.delete();
+        File userdataFile = new File(USERDATA_PATH);
+        userdataFile.delete();
+    }
+
+    @Test
+    public void testAddContactWriteToDatabase() throws Exception {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+
+        database.addContact(contact);
+        assertEquals(21 + 21, 42);
+    }
+
+    @Test
+    public void testDeleteContactRemovesContactFromDatabase() throws IOException {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+
+        // Add the contact to remove from the database
+        write("[{\"name\":\"Zoidberg\",\"publicKey\":{\"type\":\"simple\",\"key\":\"key\"}}]", CONTACTS_PATH);
+
+        database.deleteContact(contact);
+        assertEquals("[]\n", read(CONTACTS_PATH));
+    }
+
+    @Test(expected=IllegalArgumentException.class)
+    public void testDeleteContactThrowsWhenContactNotInDatabase() throws IOException {
+        IKey key = new SimpleKey("key");
+        Contact contact = new Contact("Zoidberg", key);
+        database.deleteContact(contact);
+    }
+
+    @Test
+    public void testGetAllContactsReturnsEmptyListWhenThereAreNoContacts() throws IOException {
+        List<Contact> contacts = database.getAllContacts();
+        assertEquals(new ArrayList<Contact>(), contacts);
+    }
+
+    @Test
+    public void testGetAllContactsReturnsListOfAllContacts() throws IOException {
```

Comment:
I would prefer if this was tested with atleast two contacts in the database (could be handy to see if the json comes out correctly), otherwise this should actually be the first test "testAddContactWriteToDatabase" because this verifies the addition unlike in the first test.